compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCheckNotNullCallChecker.kt

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirAnonymousFunction
+import org.jetbrains.kotlin.fir.expressions.FirCallableReferenceAccess
+import org.jetbrains.kotlin.fir.expressions.FirCheckNotNullCall
+import org.jetbrains.kotlin.fir.expressions.FirResolvedQualifier
+import org.jetbrains.kotlin.fir.expressions.FirStatement
+import org.jetbrains.kotlin.fir.types.ConeKotlinType
+import org.jetbrains.kotlin.fir.types.canBeNull
+import org.jetbrains.kotlin.fir.types.coneTypeSafe
+
+object FirCheckNotNullCallChecker : FirBasicExpressionChecker() {
+    override fun check(expression: FirStatement, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (expression !is FirCheckNotNullCall) return
+
+        val argument = expression.argumentList.arguments.singleOrNull() ?: return
+        if (argument is FirAnonymousFunction && argument.isLambda) {
+            reporter.reportOn(expression.source, FirErrors.NOT_NULL_ASSERTION_ON_LAMBDA_EXPRESSION, context)
+            return
+        }
+        if (argument is FirCallableReferenceAccess) {
+            reporter.reportOn(expression.source, FirErrors.NOT_NULL_ASSERTION_ON_CALLABLE_REFERENCE, context)
+            return
+        }
+        if (argument is FirResolvedQualifier) {
```

Comment:
FE1.0 is reporting `UNNECESSARY_NOT_NULL_ASSERTION` for the following code.

```
fun foo() {
    Foo!!.let {  }
}

object Foo {

}
```

Reply:
Thanks for finding that!

So, I reverted here and double-checked why I added this, and found what I wanted to avoid was actually resolved yet untyped(!) qualifier (e.g., `System`). That's handled like this: https://github.com/JetBrains/kotlin/blob/master/compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/ResolveUtils.kt#L147

I left a comment just in case the corresponding part in resolution is changed.